### PR TITLE
[FIX] DecoyDatabase detects already decoy'd DBs

### LINKED ISSUE #4917: DecoyDatabase should detect already decoy'd DBs
our example FASTA files in `share` sometimes contain outdated Decoy symbols (e.g. `_rev` as suffix).
If one is not careful, this can lead to subtle errors which are a bit tricky to find when a downstream tool crashes.

At least our DecoyDatabase tool should check if its consuming a decoy'ed DB (the above example is currently happily reversed **again**).

This is already mentioned in  #4662, but I think its very important to fix and to prevent accidental errors with high potential for frustration.

### PR DESCRIPTION
**Pull request by hbeschorner**
# Description

Added check, so DecoyDatabase will terminate if input data is already decoy’d.
Also added tests for the function.

Fixes #4917

# Checklist:
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- `/reformat` (experimental) applies the clang-format style changes as additional commit
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds

### CODE REVIEW

**Comment by cbielow** [comment]
I'm biased for review because these are my students.
Review by @jpfeuffer or @timosachsenberg maybe? :)

File: src/openms/include/OpenMS/DATASTRUCTURES/FASTAContainer.h [hunk 1]
```diff
@@ -360,141 +372,150 @@ class DecoyHelper
...
-  
+
+  /**
+  @brief Function to count the occurrences of decoy strings in a given set of protein names
+
+  For tested decoy strings see DecoyHelper::affixes.
+  Returns all data needed for interpretation (see DecoyHelper::DecoyStatistics).
+  */
+  template<typename T>
+  static DecoyStatistics countDecoys(FASTAContainer<T>& proteins)
```
Comment (cbielow) [inline src/openms/include/OpenMS/DATASTRUCTURES/FASTAContainer.h:456]: one thing I just realized though: a class test for this function would be nice :)
add it here: https://github.com/cbielow/OpenMS/blob/develop/src/tests/class_tests/openms/source/FASTAContainer_test.cpp#L61

**Review by timosachsenberg — APPROVED** [review]
Nice first PR. Looks very clean.
As chris pointed out, a class test for countDecoys would be nice.

**Comment by jpfeuffer** [comment]
How about allowing other decoy strings with the force parameter. If you want, undocumented. Just in case someone, even one of us, suddenly has a weird requirement. E.g. for entrapment decoys, or if you want multiple decoys for one peptide to make the FDR calculations more stable.

**Comment by timosachsenberg** [comment]
But would you run the tool several times? Do you have a use care in mind where this would work?

**Comment by jpfeuffer** [comment]
No, the tool is too inflexible for this anyway. Forcing a set of strings makes it even worse imho.

**Comment by jpfeuffer** [comment]
If this is a needed in a pipeline for example, the development of the pipeline stalls for 6+ months until the next openms release.

**Comment by timosachsenberg** [comment]
ah you mean just skipping the test if force flag is provided? Yeah that makes sense

File: src/openms/include/OpenMS/DATASTRUCTURES/FASTAContainer.h (hunk 1 above)
Comment (vrossow) [reply src/openms/include/OpenMS/DATASTRUCTURES/FASTAContainer.h:472]: Class tests for countDecoys and and findDecoyString are now added.

**Comment by hbeschorner** [comment]
Added force flag, so all custom decoy strings should be possible (warning if the custom decoy string will not be detected in check for decoys)

**Review by jpfeuffer — APPROVED** [review]

**Pull request merged**